### PR TITLE
Replaced Creation Floating Action Buttons for Events

### PR DESCRIPTION
**Pull request by Bonsai8863**
### Proposed changes

- Replaced FAB for creation of Incidents, Sightings, and Observed data.

### Checklist

- [x] I consider the submitted work as finished
- [x] I tested the code for its functionality
- [x] I wrote test cases for the relevant uses case (coverage and e2e)
- [ ] I added/update the relevant documentation (either on github or on notion)
- [x] Where necessary I refactored code to improve the overall quality

### CODE REVIEW

File: opencti-platform/opencti-front/src/private/components/events/stix_sighting_relationships/EntityStixSightingRelationships.tsx [hunk 1]
```diff
@@ -186,39 +250,15 @@ const EntityStixSightingRelationships: FunctionComponent<EntityStixSightingRelat
...
-      </Security>
+      {!FABReplaced
+        && <SightingCreationComponent
+          isTo={isTo}
+          entityId={entityId}
+          noPadding={noPadding}
+          paginationOptions={finalPaginationOptions}
+          stixCoreObjectTypes={stixCoreObjectTypes}
+           />
+      }
```
Comment (labo-flg) [inline opencti-platform/opencti-front/src/private/components/events/stix_sighting_relationships/EntityStixSightingRelationships.tsx]: We prefer this formatting for consistency (but I see our linter is not properly configured for JSX parts).

```suggestion
      {!FABReplaced && (
        <SightingCreationComponent
          isTo={isTo}
          entityId={entityId}
          noPadding={noPadding}
          paginationOptions={finalPaginationOptions}
          stixCoreObjectTypes={stixCoreObjectTypes}
        />
      )}
```

File: opencti-platform/opencti-front/src/private/components/events/stix_sighting_relationships/EntityStixSightingRelationships.tsx [hunk 2]
```diff
@@ -154,6 +210,14 @@ const EntityStixSightingRelationships: FunctionComponent<EntityStixSightingRelat
...
           paginationOptions={finalPaginationOptions}
           numberOfElements={numberOfElements}
+          createButton={FABReplaced && <SightingCreationComponent
+            isTo={isTo}
+            entityId={entityId}
+            noPadding={noPadding}
+            paginationOptions={finalPaginationOptions}
+            stixCoreObjectTypes={stixCoreObjectTypes}
+            variant='controlledDial'
+                                       />}
```
Comment (labo-flg) [inline opencti-platform/opencti-front/src/private/components/events/stix_sighting_relationships/EntityStixSightingRelationships.tsx]:
```suggestion
          createButton={FABReplaced && (
            <SightingCreationComponent
              isTo={isTo}
              entityId={entityId}
              noPadding={noPadding}
              paginationOptions={finalPaginationOptions}
              stixCoreObjectTypes={stixCoreObjectTypes}
              variant='controlledDial'
            />
          )}
```

File: opencti-platform/opencti-front/src/private/components/events/stix_sighting_relationships/EntityStixSightingRelationships.tsx [hunk 3]
```diff
@@ -25,6 +26,58 @@ const useStyles = makeStyles(() => ({
   },
 }));
 
+interface SightingCreationComponentProps {
+  isTo: boolean,
+  entityId: string,
+  noPadding?: boolean,
+  paginationOptions: EntityStixSightingRelationshipsLinesPaginationQuery$variables
+  stixCoreObjectTypes: string[],
+  variant?: string,
```
Comment (labo-flg) [inline opencti-platform/opencti-front/src/private/components/events/stix_sighting_relationships/EntityStixSightingRelationships.tsx]: Could we have a fixed list please ?
it's more an enum than a free string.

File: opencti-platform/opencti-front/src/private/components/events/Incidents.tsx [hunk 4]
```diff
@@ -135,6 +138,9 @@ const Incidents: FunctionComponent = () => {
           paginationOptions={queryPaginationOptions}
           numberOfElements={numberOfElements}
           iconExtension={true}
+          createButton={FABReplaced && <Security needs={[KNOWLEDGE_KNUPDATE]}>
+            <IncidentCreation paginationOptions={queryPaginationOptions} />
+          </Security>}
```
Comment (labo-flg) [inline opencti-platform/opencti-front/src/private/components/events/Incidents.tsx]:
```suggestion
          createButton={FABReplaced && (
            <Security needs={[KNOWLEDGE_KNUPDATE]}>
              <IncidentCreation paginationOptions={queryPaginationOptions} />
            </Security>
          )}
```

File: opencti-platform/opencti-front/src/private/components/events/Incidents.tsx [hunk 5]
```diff
@@ -179,9 +185,11 @@ const Incidents: FunctionComponent = () => {
...
       <Breadcrumbs variant="list" elements={[{ label: t_i18n('Events') }, { label: t_i18n('Incidents'), current: true }]} />
       {renderLines()}
-      <Security needs={[KNOWLEDGE_KNUPDATE]}>
-        <IncidentCreation paginationOptions={queryPaginationOptions} />
-      </Security>
+      {!FABReplaced
+        && <Security needs={[KNOWLEDGE_KNUPDATE]}>
+          <IncidentCreation paginationOptions={queryPaginationOptions} />
+        </Security>
+      }
```
Comment (labo-flg) [inline opencti-platform/opencti-front/src/private/components/events/Incidents.tsx]:
```suggestion
      {!FABReplaced && (
        <Security needs={[KNOWLEDGE_KNUPDATE]}>
          <IncidentCreation paginationOptions={queryPaginationOptions} />
        </Security>
      )}
```

File: opencti-platform/opencti-front/src/private/components/events/ObservedDatas.tsx [hunk 6]
```diff
@@ -119,6 +123,9 @@ const ObservedDatas: FunctionComponent = () => {
           paginationOptions={queryPaginationOptions}
           numberOfElements={numberOfElements}
           iconExtension={true}
+          createButton={FABReplaced && <Security needs={[KNOWLEDGE_KNUPDATE]}>
+            <ObservedDataCreation paginationOptions={queryPaginationOptions}/>
+          </Security>}
```
Comment (labo-flg) [inline opencti-platform/opencti-front/src/private/components/events/ObservedDatas.tsx]:
```suggestion
          createButton={FABReplaced && (
            <Security needs={[KNOWLEDGE_KNUPDATE]}>
              <ObservedDataCreation paginationOptions={queryPaginationOptions}/>
            </Security>
          )}
```

File: opencti-platform/opencti-front/src/private/components/events/ObservedDatas.tsx [hunk 7]
```diff
@@ -163,9 +170,11 @@ const ObservedDatas: FunctionComponent = () => {
...
           <Breadcrumbs variant="list" elements={[{ label: t_i18n('Events') }, { label: t_i18n('Observed datas'), current: true }]} />
           {renderLines(platformModuleHelpers)}
-          <Security needs={[KNOWLEDGE_KNUPDATE]}>
-            <ObservedDataCreation paginationOptions={queryPaginationOptions}/>
-          </Security>
+          {!FABReplaced
+            && <Security needs={[KNOWLEDGE_KNUPDATE]}>
+              <ObservedDataCreation paginationOptions={queryPaginationOptions}/>
+            </Security>
+          }
```
Comment (labo-flg) [inline opencti-platform/opencti-front/src/private/components/events/ObservedDatas.tsx]:
```suggestion
          {!FABReplaced && (
            <Security needs={[KNOWLEDGE_KNUPDATE]}>
              <ObservedDataCreation paginationOptions={queryPaginationOptions}/>
            </Security>
          )}
```

File: opencti-platform/opencti-front/src/private/components/events/incidents/IncidentCreation.tsx [hunk 8]
```diff
@@ -284,11 +290,14 @@ const IncidentCreation = ({
...
   const { t_i18n } = useFormatter();
+  const { isFeatureEnable } = useHelper();
   const updater = (store: RecordSourceSelectorProxy) => insertNode(store, 'Pagination_incidents', paginationOptions, 'incidentAdd');
+  const FABReplaced = isFeatureEnable('FAB_REPLACEMENT');
   return (
     <Drawer
       title={t_i18n('Create an incident')}
-      variant={DrawerVariant.create}
+      variant={FABReplaced ? undefined : DrawerVariant.create}
+      controlledDial={FABReplaced ? CreateEntityControlledDial('entity_Incident') : undefined}
```
Comment (labo-flg) [inline opencti-platform/opencti-front/src/private/components/events/incidents/IncidentCreation.tsx]: I know this has been already merged, but I'm really not fond of this factory CreateEntityControlledDial.

IMO a more "react" solution would be to have a component `EntityControlledDial` that has the entityType in props:

```ts
import React from 'react';
import { Button } from '@mui/material';
import { useFormatter } from 'src/components/i18n';

interface EntityControlledDialProps {
  entityType: string
  onOpen: () => unknown
}

const EntityControlledDial: React.FC<EntityControlledDialProps> = ({ onOpen, entityType }) => {
  const { t_i18n } = useFormatter();
  const buttonValue = `${t_i18n('Create')} ${t_i18n(entityType)}`;
  return (
    <Button
      onClick={onOpen}
      color='primary'
      size='small'
      variant='contained'
      aria-label={buttonValue}
      title={buttonValue}
      sx={{
        marginLeft: '10px',
        padding: '7px 10px',
      }}
    >
      <div style={{ display: 'flex' }}>
        {buttonValue}
      </div>
    </Button>
  );
};

export default EntityControlledDial;
```

And then define specific ControlledDials where we need to. For instance here in IncidentCreation.tsx

```ts
//...
const IncidentControlledDial = (props: Omit<EntityControlledDialProps, 'entityType'>) => (
  <EntityControlledDial entityType={'entity_Incident'} {...props} />
);
//...
  <Drawer
      title={t_i18n('Create an incident')}
      variant={FABReplaced ? undefined : DrawerVariant.create}
      controlledDial={FABReplaced ? IncidentControlledDial : undefined}
    >

```

File: opencti-platform/opencti-front/src/private/components/events/stix_sighting_relationships/StixSightingRelationshipCreationFromEntity.jsx [hunk 9]
```diff
@@ -455,29 +460,41 @@ const StixSightingRelationshipCreationFromEntity = ({
       </div>
     );
   };
+
+  const openElement = () => {
+    switch (variant) {
+      case 'controlledDial':
+        return CreateEntityControlledDial('entity_Sighting')({ onOpen: handleOpen });
```
Comment (labo-flg) [inline opencti-platform/opencti-front/src/private/components/events/stix_sighting_relationships/StixSightingRelationshipCreationFromEntity.jsx]: This is why I'm not fond of factories ; it makes the code unclear.
We can have a dedicated component like I said earlier.

**Comment by labo-flg** [comment]
Functional tests are all OK ; I've added some comments. [This one](https://github.com/OpenCTI-Platform/opencti/pull/7256#discussion_r1651101931) is the one that truly matters.

File: opencti-platform/opencti-front/src/private/components/events/Incidents.tsx [hunk 10]
```diff
@@ -111,6 +113,7 @@ const Incidents: FunctionComponent = () => {
       isSortable: isRuntimeSort,
     },
   };
+  const FABReplaced = isFeatureEnable('FAB_REPLACEMENT');
```
Comment (labo-flg) [inline opencti-platform/opencti-front/src/private/components/events/Incidents.tsx]:
```suggestion
  const isFABReplaced = isFeatureEnable('FAB_REPLACEMENT');
```

by convention, simple variables do not start with a capital ; and for clarity I would advise to start boolean vars with "is" for clarity.

File: opencti-platform/opencti-front/src/private/components/events/stix_sighting_relationships/EntityStixSightingRelationships.tsx (hunk 3 above)
Comment (Bonsai8863) [reply opencti-platform/opencti-front/src/private/components/events/stix_sighting_relationships/EntityStixSightingRelationships.tsx]: For variant? Any action on the other three lines here?

Comment (Bonsai8863) [reply opencti-platform/opencti-front/src/private/components/events/stix_sighting_relationships/EntityStixSightingRelationships.tsx]:
```suggestion
  variant?: 'controlledDial' | 'inLine',
```
Is this good?

File: opencti-platform/opencti-front/src/components/CreateEntityControlledDial.tsx [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+const CreateEntityControlledDial: FunctionComponent<CreateEntityControlledDialProps> = ({
+  onOpen,
+  entityType,
+  color = 'primary',
+  size = 'small',
+  variant = 'contained',
+  style,
+}) => {
+  const { t_i18n } = useFormatter();
+  const buttonValue = `${t_i18n('Create')} ${t_i18n(entityType)}`;
```
Comment (labo-flg) [inline opencti-platform/opencti-front/src/components/CreateEntityControlledDial.tsx]: I suggest we get rid of the `entity_` prefix in all calls to `<CreateEntityControlledDial />`.
This way the prop `entityType` is _really_ an entity type, and the prefix tweak stays for i18n only

```suggestion
  const buttonValue = `${t_i18n('Create')} ${t_i18n(`entity_${entityType}`)}`;
```

Comment (labo-flg) [reply opencti-platform/opencti-front/src/components/CreateEntityControlledDial.tsx]: translating "Create X" as 2 separate calls to `t_i18n` would force the translation order.
In German for instance, the 2 words would actually be inversed.

One way to solve this is to use variables in the intl token.

```suggestion
  const buttonValue = t_i18n('', {
    id: 'Create ...',
    values: { entity_type: t_i18n(`entity_${entityType}`) },
  });
```

and in the translation files

```
"Create ...": "Create {entity_type}",
```

File: opencti-platform/opencti-front/src/private/components/events/stix_sighting_relationships/EntityStixSightingRelationships.tsx (hunk 3 above)
Comment (labo-flg) [reply opencti-platform/opencti-front/src/private/components/events/stix_sighting_relationships/EntityStixSightingRelationships.tsx]: sorry I missed this comment. Yes it's good!

**Review by labo-flg — APPROVED** [review]
All seems good, tested OK !
Thanks, nice job :)

**Pull request merged**